LFS-733: Improve month-year handling in date question

acrowthe opened the pull request:
- Add detection for yyyy-MM and MM-yyyy date formats
- Convert handling of said formats from "month" input to regexp backed text field
  - limit month to number of format 01 through 12 and year to 4 digits
- Store as regular date type by appending day value of 01

Testing:
A testing branch has been created [here](https://github.com/ccmbioinfo/lfs/tree/LFS-733-test) with multiple different variations of date questions added to the top of the Demographics form.

Previous testing instructions:
Create a date question with:
- A `dateFormat` of `yyyy-MM` or `MM-yyyy`
- A minAnswers and maxAnswers filled out, either 0 or 1.

This should result in a text field which prevents invalid strings from being entered, limiting user input to the format `####/##`, where the last 2 numbers are in the range 01 to 12.

acrowthe committed:
LFS-733: Improve month-year handling in date question
- Add detection for yyyy-MM and MM-yyyy date formats
- Convert handling of said formats from "month" input to regexp backed text field
  - limit month to number of format 01 through 12 and year to 4 digits
- Store as regular date type by appending day value of 01

marta- commented on modules/data-entry/src/main/frontend/src/dataHomepage/FilterComponents/DateFilter.jsx:
```diff
@@ -23,7 +23,7 @@ import PropTypes from "prop-types";
 import FilterComponentManager from "./FilterComponentManager.jsx";
 import { DEFAULT_COMPARATORS, UNARY_COMPARATORS, VALUE_COMPARATORS } from "./FilterComparators.jsx";
-import { DATE_FORMATS, MONTH_FORMAT } from "../../questionnaire/DateQuestion.jsx";
+import { DATE_FORMATS, MONTH_FORMAT, MONTH_FORMATS } from "../../questionnaire/DateQuestion.jsx";
```
Shouldn't `MONTH_FORMAT` be removed?
```suggestion
import { DATE_FORMATS, MONTH_FORMATS } from "../../questionnaire/DateQuestion.jsx";
```

sashaandjic commented:
Once I enter the correct format, the field is flagged as INCOMPLETE,INVALD.
![image](https://user-images.githubusercontent.com/23243244/100280269-e4016180-2f35-11eb-8867-7596f150c8ae.png)

acrowthe committed:
LFS-733: Improve month-year handling in date question
Remove now invalid import

acrowthe commented:
@sashaandjic Looks like minAnswers and maxAnswers have to be filled out or the field will always get marked as invalid/incomplete by the backend validation. I have updated the testing instructions

sashaandjic commented:
> @sashaandjic Looks like minAnswers and maxAnswers have to be filled out or the field will always get marked as invalid/incomplete by the backend validation. I have updated the testing instructions

Is it out of scope of this PR?

acrowthe commented:
> > @sashaandjic Looks like minAnswers and maxAnswers have to be filled out or the field will always get marked as invalid/incomplete by the backend validation. I have updated the testing instructions
> 
> @acrowthe Is it out of scope of this PR?

Yes it is out of scope. I'm working on reproduction steps on dev to create issues.

sashaandjic reviewed (approved):
tested

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaire/DateQuestion.jsx:
```diff
@@ -134,6 +142,11 @@ function DateQuestion(props) {
+  if (isMonth) {
+    displayFormat = "yyyy/MM/dd";
+    dateFormat = "yyyy-MM-dd";
+  }
```
What is this meant to do?

acrowthe replied:
Added explanatory comments to this section

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaire/DateQuestion.jsx:
```diff
@@ -164,16 +177,44 @@ function DateQuestion(props) {
+  let validateMonthString = (value) => {
+    if (!monthYearTest.test(value)) {
+      setError(true);
+      setErrorMessage("Please enter a month in the format yyyy/MM, such as 2000/01");
```
`yyyy/MM` is too technical (regular users may be confused by the uppercase MM.

It is also too hardcoded and doesn't reflect the configured date format.

sdumitriu reviewed (changes requested):
I see a few problems with this PR:
- although the MM-yyyy format is supposed to be supported, it is not
- although the format is supposed to be yyyy-MM, the only value format supported is yyyy/MM (with a slash)
- typing 2012/02 and then another keystroke erases the field; same for a 5-digit year, typing 10212/03 erases the value
- I would like `2012/3` (without a leading 0) to be supported as well
- isn't there any browser native widget for picking yyyy-MM dates?
- although the display format is hardcoded to yyyy/MM, actually entering that format (with /) causes the input to not accept any value at all
- the empty field gives no indication of what the user is supposed to enter in it; the first hint comes when starting to type and an error validation message shows up. This is not intuitive (the format should be set as a placeholder in the input), and scary (the error should only show up when the user has finished typing the value and left the field).

acrowthe commented:
> * although the format is supposed to be yyyy-MM, the only value format supported is yyyy/MM (with a slash)

Sling expects date strings to use dashes, in the format `yyyy-mm-dd`. User input was specified to use slashes for the questionnaire and the browser date picker also uses slashes. Thus, there is some disparity between front end formats and back end formats.

> * isn't there any browser native widget for picking yyyy-MM dates?

There is an input with type="month", but it is not supported by Firefox or Safari ([caniuse link](https://caniuse.com/mdn-html_elements_input_input-month))

> * the first hint comes when starting to type and an error validation message shows up. This is ... scary (the error should only show up when the user has finished typing the value and left the field).

I believe this is the same behavior as the regexp support for regular text questions. I can change it here if preferred though.

The other issues I'll look into on Monday

acrowthe committed:
LFS-733: Improve month-year handling in date question
- Add support for multiple month-year formats
- Add placeholder text to month-year formats
Fix type interval date questions
- Previously, the timestamp default was always being used
Fix limits
- Limits were being ignored

acrowthe commented:
Added support for both yyyy/mm and mm/yyyy date questions.
Added ability to enter months without a leading 0
Changed field validation to occur when the user switches focus away from the input.
Added a placeholder describing entry format.
Fixed issues where field would stop taking input, and where field would be unexpectedly cleared.
